### PR TITLE
Add `u-svg-inline-bg` mixin for inline SVG background

### PR DESCRIPTION
**Pull request by anselmbradford**
## Additions
- Adds `.u-svg-inline-bg( @name )` mixin and associated plugin.

## Removals
- Remove cf-core dependency from cf-icons.
- Removes `@cf-icon-path` variable.

## Changes
- Convert cf-tables to use SVG image instead of PNG for sort button.
- Adds missing "font variables" entry from cf-typography TOC.

## Testing

1. `gulp build && gulp docs && bundle exec jekyll serve` should pass and launch the docs.
2. Visit http://localhost:4000/components/cf-tables/#sortable-tables and compare to https://cfpb.github.io/capital-framework/components/cf-tables/#sortable-tables
3. Visit http://localhost:4000/components/cf-forms/#select-dropdown and compare to https://cfpb.github.io/capital-framework/components/cf-forms/#select-dropdown
4. Visit http://localhost:4000/components/cf-forms/#checkboxes-and-radio-inputs and compare to https://cfpb.github.io/capital-framework/components/cf-forms/#checkboxes-and-radio-inputs

## Screenshots

Before:
Screen Shot 2019-07-09 at 11 22 34 AM

After:
Screen Shot 2019-07-09 at 11 32 08 AM

## Notes

- We may want to scrutinize whether packages other than cf-icons also do not depend on cf-core and remove those dependencies.

### CODE REVIEW

**Comment by Scotchester** [comment]
Buttons and Pagination use icons in buttons and Expendables use the up and
down chevrons in the Show/Hide label.

>

**Comment by anselmbradford** [comment]
> Buttons and Pagination use icons in buttons and Expendables use the up and
down chevrons in the Show/Hide label.

So I wasn't thinking clearly here in respect to cf-buttons etc being installed in isolation, since there are no direct CSS dependencies within those packages, but rather depend on the convention of adding SVG markup within the component with the `cf-icon-svg` class. However, thinking about this, we have some inconsistencies:

 - Components that depend on cf-icon do so because by convention their markup should include an SVG icon. However, some packages, like [cf-tables do not depend on cf-icons](https://github.com/cfpb/capital-framework/blob/master/packages/cf-tables/package.json), but _could_ include SVG icons in the cells of the table. If it's the developer's responsibility to include cf-icons if they have SVG icon content within cf-tables content area, would there also be an argument that it is the developer's responsibility to do so for all packages that include SVG icon content (so cf-buttons, cf-expandables, and cf-pagination)?

- cf-tables [uses a data URL to include a PNG](https://github.com/cfpb/capital-framework/blob/master/packages/cf-tables/src/cf-tables.less#L90) for its chevron. First, [this should be an SVG](https://github.com/cfpb/capital-framework/issues/940), and second, if this is an SVG, why couldn't all SVG icons be included this way and cf-icons only be used as a dependency for direct icons used outside of another component's pattern.

- cf-forms [does include SVG as a `background-image`](https://github.com/cfpb/capital-framework/blob/master/packages/cf-forms/src/atoms/select.less#L67), which requires making the SVG file available to the CSS, such as is done [here](https://github.com/cfpb/capital-framework/tree/master/docs/dist/icons) in the docs, and requires the use of the `@cf-icon-path`. What if this was instead an inlined SVG image set on the background-image?

Proposal: What if we converted all icons used within CF components to be inlined SVG in the CSS as data URLs. Here's a commit of what that would look like https://github.com/cfpb/capital-framework/pull/938/commits/be50dbe4e8d3771b079fccf3c68fcd984dd78e6e
If we wanted to reference the cf-icon source file, what if we used something similar to https://github.com/atlassian/less-plugin-inline-svg (or a simpler find and replace) in the build step to embed the SVG. The path in the source file would be the default backup, if the developer used a build process that didn't replace them. Thoughts?


this makes me think, if we're talking about these packages in isolation, should cf-icons be a dependency of all? Take for example, if cf-tables were to be installed in isolation. [Cf-tables does not depend on cf-icons](https://github.com/cfpb/capital-framework/blob/master/packages/cf-tables/package.json), but an icon could be used inside the table cells. This would require including cf-icons. But how is this

**Comment by sephcoster** [comment]
Unsure if this would all work given LESS and how it handles variables, but if we go this route could we create the `data:image/svg+xml;charset=UTF-8,<svg xmlns=` strings as @ variables and call them that way to make it easier to reference?

End result might be something like: `background: url ( @icon-down-embed )`which would reference the embedded SVG.

**Comment by anselmbradford** [comment]
Okay I've totally re-worked this PR and rebased and force pushed, so throw out what you have and re-pull if checking this again.

I realized we could use a custom less plugin to run JavaScript at build-time to inject the source SVG files inline into the background-image property via a mixin, which this PR adds.

Icons are added two ways: in the markup + CSS styling, and purely via CSS as a background-image. 
Say all icons were delivered purely via CSS. This offers some interesting trade-offs to consider:

- `+/-`The inline SVG moves from the markup to the CSS, making the markup smaller and the CSS larger. 
   - `+` This makes the dom-tree node count lower, possibly making time to first content quicker.
   - `+` This makes the icons absent when CSS is not loaded, versus unstyled and huge (https://github.com/cfpb/capital-framework/issues/863)
   - `-` Since cf-icons is delivered to all pages, all embedded SVGs get delivered to all pages. If this was ALL icons, that's 1.6MB (however, this may be cached after first load). This could possibly be addressed through further work in https://github.com/cfpb/cfgov-refresh/issues/253.
- `-` The width of the icon becomes more difficult. As a background-image the width needs to be explicitly set, which varies by icon (icons have consistent height, and non-consistent widths).
- `+` The bewd [`svg_icon`](https://github.com/cfpb/cfgov-refresh/blob/b875f08b980f1acea53d8f37a7d3aa0d5d6f2e1f/cfgov/agreements/jinja2/agreements/index.html#L59) function could be eliminated if all icons were delivered via CSS.
- `-` Styling the icons becomes more difficult, since they aren't included in the dom-tree any longer. This means CSS [`filters`](https://css-tricks.com/solved-with-css-colorizing-svg-backgrounds/) need to be used to colorize them. The source icons would need to be set to white instead of black for the brightness filter to work.

**Comment by niqjohnson** [comment]
@anselmbradford: @caheberer and I took a look at this, and we're good replacing the custom up/down filled triangles with our standard up/down chevrons in sortable tables. Turns out the reason we didn't go with the chevrons originally was because the old chevrons were thinner and didn't have enough contrast in table headers. The recent re-draw of the icons fixes that issue, so it makes sense to get rid of that custom triangle icon.

One issue we did notice is that the current chevron is sitting a little low in the cell, which is causing a couple of minor visual issues:

- The chevron looks off-center vertically in the cell
- Sortable cells have a slightly lower baseline than non-sortable cells

Can you vertically center the chevron with the cell's text? I did a quick mockup of that in Photoshop by pushing the chevron (both up and down states) up by 3px to get everything aligned (I added guides to the images to make it a little easier to eyeball things):

State | Without guides | With guides
---- | ---- | ----
Current (down) | current | current-guides
Pushed up 3px (down) | new | new-guides
Pushed up 3px (up) | flipped | flipped-guides

Once that alignment is fixed, this change looks a-ok from a design and UX perspective. Thanks!

File: packages/cf-icons/src/cf-icons.less [hunk 1]
```diff
@@ -11,31 +11,46 @@
...
 
+//
+// Embedded inline SVG data URLs.
+//
+
+// Mixin to inject an SVG from the ./icons/ directory
+// into a background-image property.
+@plugin "cf-icons-svg-inline";
+.u-svg-inline-bg( @type ) {
+    @svg: cf-icons-svg-inline( @type );
```
Comment (Scotchester) [inline packages/cf-icons/src/cf-icons.less]: Trivial, but `@type` seems like a slightly odd choice for this parameter. I think `@name` would be a little more intuitive.

Comment (anselmbradford) [reply packages/cf-icons/src/cf-icons.less]: Changed in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-icons/src/cf-icons.less [hunk 2]
```diff
@@ -11,31 +11,46 @@
 // Size variables
 //
 
-// Icon height matches the 19px rendered canvas of text set in Avenir Next
-// sized at 16px ( 19/16 = 1.1875 )
+// Icon SVGs viewbox is 1200 (h) x 1000 (w).
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-icons/src/cf-icons.less]: Changed in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-forms/src/molecules/form-fields.less [hunk 3]
```diff
@@ -121,10 +121,7 @@
             }
 
             &:checked + .a-label:before {
-                background-image: url( '@{cf-icon-path}/approved.svg' );
-                background-size: auto @cf-icon-height;
-                background-repeat: no-repeat;
-                background-position: center 0;
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-forms/src/molecules/form-fields.less]: Changed in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-icons/src/cf-icons-svg-inline.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+const path = require( 'path' );
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-icons/src/cf-icons-svg-inline.js]: Added in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-icons/src/cf-icons-svg-inline.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+module.exports = {
+  install: function( less, pluginManager, functions ) {
+    functions.add( 'cf-icons-svg-inline', svgName => {
+      // Retrieve this plugin scripts path so we can fake __dirname.
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-icons/src/cf-icons-svg-inline.js]: Fixed in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-icons/src/cf-icons-svg-inline.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+      const svg = less.fs.readFileSync(
+        path.join( __dirname, `./icons/${ svgName.value }.svg` ),
+        'utf8'
+      );
+
+      return svg;
+    } );
+  }
+}
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-icons/src/cf-icons-svg-inline.js]: Fixed in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-tables/src/cf-tables.less [hunk 7]
```diff
@@ -22,6 +22,7 @@
 //
 
 @import '../../cf-core/src/cf-core.less';
+@import '../../cf-icons/src/cf-icons.less';
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-tables/src/cf-tables.less]: Added in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

File: packages/cf-icons/src/cf-icons.less [hunk 8]
```diff
@@ -11,31 +11,46 @@
...
     .lt-ie10 & {
-        // IE 9 require a width otherwise the SVG takes up 100%
+        // IE 9 require a width otherwise the SVG takes up 100%.
         width: 1em;
     }
 
     .lt-ie9 & {
-        // IE 8 doesn't support currentColor, hide icons and let the paired
-        // text stand on its own
+        // IE 8 doesn't support currentColor,
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply packages/cf-icons/src/cf-icons.less]: Fixed in https://github.com/cfpb/capital-framework/pull/938/commits/f91b0f5f1615028ebee714fb637193a38ba868ce

**Comment by anselmbradford** [comment]
@Scotchester @niqjohnson Ready for review again. Thanks!

**Comment by niqjohnson** [comment]
@anselmbradford: The new alignment in the sortable tables looks perfect to me—thanks. I took a quick look at the other changes, too, and everything looked the same locally as in production (as expected). So this looks 👍 from the design and UX sides.

**Pull request merged**